### PR TITLE
[Shopify][BC Idea]: Enhancing Shopify Connector for Accurate Price Synchronization with Business Central based on Customer

### LINKED ISSUE #26980: [BC Idea]: Enhancing Shopify Connector for Accurate Price Synchronization with Business Central based on Customer
### BC Idea Link

https://experience.dynamics.com/ideas/idea/?ideaid=34dcafc6-5c57-ef11-b4ac-6045bd8476a6

### Description

Currently, there is a need to synchronize prices between Business Central and Shopify, where prices are determined by the lowest pricing logic applied in Business Central. This involves complex pricing scenarios like customer-specific discounts, item discount groups, and other discount mechanisms. However, the existing Shopify Connector does not account for these complexities, leading to potential discrepancies between the prices in Business Central and those displayed in Shopify.

### I will provide the implementation for this BC Idea

- [X] I will provide the implementation for this BC Idea
Internal work item: [AB#544010](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/544010)

### PR DESCRIPTION
**Pull request by VijayErpEff**
#### Summary 
Added Customer No to the 
          table 30152 "Shpfy Catalog"
          page 30159 "Shpfy Catalogs"

Added logic to use customer specific pricing from catalog
           codeunit 30182 "Shpfy Product Price Calc."

Fixes #26980
Fixes [AB#544010](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/544010)

### CODE REVIEW

**Comment by VijayErpEff** [comment]
@microsoft-github-policy-service agree company="ERP Efficiency Experts, LLC"

File: Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductPriceCalc.Codeunit.al [hunk 1]
```diff
@@ -91,22 +92,37 @@ codeunit 30182 "Shpfy Product Price Calc."
...
-        TempSalesHeader."Sell-to Customer No." := Shop.Code;
-        TempSalesHeader."Bill-to Customer No." := Shop.Code;
+        if CustomerNo <> '' then begin
+            if CustomerRec.GET(CustomerNo) then begin
+                TempSalesHeader."Sell-to Customer No." := CustomerNo;
+                TempSalesHeader."Bill-to Customer No." := CustomerNo;
+                TempSalesHeader."Customer Price Group" := CustomerRec."Customer Price Group";
+                TempSalesHeader."Customer Disc. Group" := CustomerRec."Customer Disc. Group";
+                TempSalesHeader."Allow Line Disc." := CustomerRec."Allow Line Disc.";
+                TempSalesHeader."Invoice Disc. Code" := CustomerRec."Invoice Disc. Code";
```
Comment (AndreiPanko) [inline Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductPriceCalc.Codeunit.al:108]: not used, as far as i know.

File: Apps/W1/Shopify/app/src/Catalogs/Pages/ShpfyCatalogs.Page.al [hunk 2]
```diff
@@ -25,6 +25,11 @@ page 30159 "Shpfy Catalogs"
                     ToolTip = 'Specifies the unique identifier for the catalog in Shopify.';
                     Editable = false;
                 }
+                field("Customer No."; Rec."Customer No.")
```
Comment (AndreiPanko) [inline Apps/W1/Shopify/app/src/Catalogs/Pages/ShpfyCatalogs.Page.al:28]: Visible = false, please.

Extend tooltip, to highlight that parameters defined on the customer card, such as Cust Disc Group, Cust Price Group, Allow Line Disc have priority over values defined in the catalog.

File: Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductPriceCalc.Codeunit.al [hunk 3]
```diff
@@ -91,22 +92,37 @@ codeunit 30182 "Shpfy Product Price Calc."
...
+        else begin
+            TempSalesHeader."Sell-to Customer No." := Shop.Code;
+            TempSalesHeader."Bill-to Customer No." := Shop.Code;
+            TempSalesHeader."Customer Price Group" := CustomerPriceGroup;
+            TempSalesHeader."Customer Disc. Group" := CustomerDiscGroup;
+            TempSalesHeader."Allow Line Disc." := AllowLineDisc;
+        end;
+
         TempSalesHeader."Gen. Bus. Posting Group" := GenBusPostingGroup;
         TempSalesHeader."VAT Bus. Posting Group" := VATBusPostingGroup;
```
Comment (AndreiPanko) [inline Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductPriceCalc.Codeunit.al:120]: Some unit test would be great.

There is a codeunit that already has a test for catalog.
https://github.com/microsoft/ALAppExtensions/blob/main/Apps/W1/Shopify/test/Catalogs/ShpfyCatalogPricesTest.Codeunit.al

I suggest to add 4 more in an equivalent manner, make sure to populate Customer No in the Catalog record.
1
CustomerX
Product1 unit price 100  Item discount category DISCOUNT
General discount for DISCOUNT= 20%
Exported prices: Product1 = 80

2
CustomerX
Product2 unit price 100  Item discount category PAINT
General discount for DISCOUNT= 20%
Customer specific discount for Product2 = 30%
Exported prices: Product2 = 70

3
CustomerX
Product3 unit price 100  Item discount category PAINT
General discount for DISCOUNT= 20%
Product specific discount for Product3 = 25% (for All Customers)
Exported prices: Product3 = 75

4
CustomerX
Product4 unit price 100  Item discount category PAINT
General discount for DISCOUNT= 20%
Customer specific discount for Product4 = 30%, Product specific discount for Product4 = 25%
Exported prices: Product4 = 70.

**Review by AndreiPanko — CHANGES_REQUESTED** [review]

File: Apps/W1/Shopify/app/src/Catalogs/Pages/ShpfyCatalogs.Page.al (hunk 2 above)
Comment (VijayErpEff) [reply Apps/W1/Shopify/app/src/Catalogs/Pages/ShpfyCatalogs.Page.al:28]: @AndreiPanko If this is marked as visible false should customer extend this app to make it visible again

File: Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductPriceCalc.Codeunit.al (hunk 3 above)
Comment (VijayErpEff) [reply Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductPriceCalc.Codeunit.al:120]: @AndreiPanko I added Test cases. Can you please take a look

File: Apps/W1/Shopify/app/src/Catalogs/Pages/ShpfyCatalogs.Page.al (hunk 2 above)
Comment (AndreiPanko) [reply Apps/W1/Shopify/app/src/Catalogs/Pages/ShpfyCatalogs.Page.al:28]: No, they will be able to do it via personalization. No development is needed.

File: Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductPriceCalc.Codeunit.al [hunk 4]
```diff
@@ -91,22 +92,37 @@ codeunit 30182 "Shpfy Product Price Calc."
     /// Create Temp Sales Header.
     /// </summary>
     local procedure CreateTempSalesHeader()
+    var
+        CustomerRec: Record Customer;
```
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductPriceCalc.Codeunit.al:96]: Why not just Customer: Record Customer; ?

File: Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductPriceCalc.Codeunit.al [hunk 5]
```diff
@@ -91,22 +92,37 @@ codeunit 30182 "Shpfy Product Price Calc."
...
     begin
         Clear(TempSalesHeader);
         TempSalesHeader."Document Type" := TempSalesHeader."Document Type"::Quote;
         TempSalesHeader."No." := Shop.Code;
-        TempSalesHeader."Sell-to Customer No." := Shop.Code;
-        TempSalesHeader."Bill-to Customer No." := Shop.Code;
+        if CustomerNo <> '' then begin
+            if CustomerRec.GET(CustomerNo) then begin
```
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductPriceCalc.Codeunit.al:102]: Customer.**Get**(CustomerNo)

File: Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductPriceCalc.Codeunit.al (hunk 1 above)
Comment (onbuyuka) [reply Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductPriceCalc.Codeunit.al:108]: yes not used, should be removed

File: Apps/W1/Shopify/test/Catalogs/ShpfyCatalogPricesTest.Codeunit.al [hunk 6]
```diff
@@ -72,4 +72,230 @@ codeunit 139646 "Shpfy Catalog Prices Test"
         // [THEN] InitPrice - InitDiscountPerc = Price
         LibraryAssert.AreNearlyEqual(InitPrice * (1 - InitDiscountPerc / 100), Price, 0.01, 'Discount Price');
     end;
+
+
```
Comment (onbuyuka) [inline Apps/W1/Shopify/test/Catalogs/ShpfyCatalogPricesTest.Codeunit.al:76]: Just one empty line should be enough

File: Apps/W1/Shopify/test/Catalogs/ShpfyCatalogPricesTest.Codeunit.al [hunk 7]
```diff
@@ -72,4 +72,230 @@ codeunit 139646 "Shpfy Catalog Prices Test"
...
+
+
+    [Test]
+    procedure UnitTestCalcCatalogPriceAllCustomers()
```
Comment (onbuyuka) [inline Apps/W1/Shopify/test/Catalogs/ShpfyCatalogPricesTest.Codeunit.al:80]: Please use [GIVEN], [WHEN], [THEN] format. You can look at other tests

File: Apps/W1/Shopify/test/Catalogs/ShpfyCatalogPricesTest.Codeunit.al [hunk 8]
```diff
@@ -72,4 +72,230 @@ codeunit 139646 "Shpfy Catalog Prices Test"
...
+        ProductPriceCalculation.CalcPrice(Item, '', '', UnitCost, Price, ComparePrice);
+
+        // Verify the results post discount application.
+        LibraryAssert.AreEqual(InitUnitCost, UnitCost, 'Unit cost should remain consistent post-discount.');
+        LibraryAssert.AreEqual(InitPrice, ComparePrice, 'Compare price should still reflect initial price pre-discount.');
+        LibraryAssert.AreNearlyEqual(InitPrice * (1 - CustDiscPerc / 100), Price, 0.01, 'Verify discounted price is calculated correctly.');
+    end;
+
+    [Test]
+    procedure UnitTestCalcCatalogPriceCustomer3()
```
Comment (onbuyuka) [inline Apps/W1/Shopify/test/Catalogs/ShpfyCatalogPricesTest.Codeunit.al:248]: It would be great if you could give a more meaningful name to this procedure

**Review by VijayErpEff — COMMENTED** [review]
@onbuyuka Can you please review Thanks

File: Apps/W1/Shopify/test/Catalogs/ShpfyCatalogPricesTest.Codeunit.al [hunk 9]
```diff
@@ -72,4 +73,241 @@ codeunit 139646 "Shpfy Catalog Prices Test"
...
+
+        // [WHEN] Recalculating prices after applying the discount.
+        ProductPriceCalculation.SetShopAndCatalog(Shop, Catalog);
+        ProductPriceCalculation.CalcPrice(Item, '', '', UnitCost, Price, ComparePrice);
+
+        // [THEN] Validate the results reflect the universal discount.
+        LibraryAssert.AreEqual(InitUnitCost, UnitCost, 'Unit cost should remain consistent after discount application.');
+        LibraryAssert.AreEqual(InitPrice, ComparePrice, 'Compare price should reflect the original price prior to any discounts.');
+        LibraryAssert.AreNearlyEqual(InitPrice * (1 - InitDiscountPerc / 100), Price, 0.01, 'The final price should accurately reflect the applied discount for all customers.');
+#endif
```
Comment (onbuyuka) [inline Apps/W1/Shopify/test/Catalogs/ShpfyCatalogPricesTest.Codeunit.al:134]: What is the reason behind this? Why do you have two different tests in one procedure?

Comment (onbuyuka) [reply Apps/W1/Shopify/test/Catalogs/ShpfyCatalogPricesTest.Codeunit.al:134]: Also why CLEAN23?

Comment (VijayErpEff) [reply Apps/W1/Shopify/test/Catalogs/ShpfyCatalogPricesTest.Codeunit.al:134]: Hello @onbuyuka I followed the original test case and use the same process as Pricelists needs to be enabled for these tests.

This was the original test

#if not CLEAN23
        ProductInitTest.CreateSalesPrice(Shop.Code, Item."No.", InitPrice);
        CustomerDiscountGroup := ProductInitTest.CreateSalesLineDiscount(Shop.Code, Item."No.", InitDiscountPerc);
#else
        CustomerDiscountGroup := ProductInitTest.CreatePriceList(Shop.Code, Item."No.", InitPrice, InitDiscountPerc);
#endif

The original test case was run without applying any discounts. The next steps involve enabling the Pricelists and applying the relevant discounts as per the process.

Comment (onbuyuka) [reply Apps/W1/Shopify/test/Catalogs/ShpfyCatalogPricesTest.Codeunit.al:134]: Okay I see, thank you for the explanation

**Review by onbuyuka — APPROVED** [review]

File: Apps/W1/Shopify/test/Catalogs/ShpfyCatalogPricesTest.Codeunit.al [hunk 10]
```diff
@@ -72,4 +73,241 @@ codeunit 139646 "Shpfy Catalog Prices Test"
...
+        LibraryAssert.AreNearlyEqual(InitPrice * (1 - CustDiscPerc / 100), Price, 0.01, 'Discounted price should be accurately calculated.');
+#endif
+    end;
+
+    [Test]
+    procedure UnitTestCalcCustomerDiscountCatalogPrice()
+    var
+        Shop: Record "Shpfy Shop";
+        Catalog: Record "Shpfy Catalog";
+        LibrarySales: Codeunit "Library - Sales";
```
Comment (JesperSchulz) [inline Apps/W1/Shopify/test/Catalogs/ShpfyCatalogPricesTest.Codeunit.al:260]: I had to make some more changes to make this pass our build system. This variable for instance is only needed with CLEAN23 tag. But it should all have been taken care of now.

Comment (VijayErpEff) [reply Apps/W1/Shopify/test/Catalogs/ShpfyCatalogPricesTest.Codeunit.al:260]: Thanks @JesperSchulz

**Review by VijayErpEff — COMMENTED** [review]
All changes complete

**Review by JesperSchulz — APPROVED** [review]

**Pull request merged**